subsys/bluetooth/audio/aics.c

```diff
@@ -0,0 +1,824 @@
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_OFFSET);
+	}
+
+	if (!len || !buf) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	/* Check opcode before length */
+	if (!VALID_AICS_OPCODE(cp->cp.opcode)) {
+		BT_DBG("Invalid opcode %u", cp->cp.opcode);
+		return BT_GATT_ERR(BT_AICS_ERR_OP_NOT_SUPPORTED);
+	}
+
+	if ((len < AICS_CP_LEN) ||
+	    (len == AICS_CP_SET_GAIN_LEN && cp->cp.opcode != BT_AICS_OPCODE_SET_GAIN) ||
+	    (len > AICS_CP_SET_GAIN_LEN)) {
+		return BT_GATT_ERR(BT_ATT_ERR_INVALID_ATTRIBUTE_LEN);
+	}
+
+	BT_DBG("Opcode %u, counter %u", cp->cp.opcode, cp->cp.counter);
+	if (cp->cp.counter != inst->state.change_counter) {
+		return BT_GATT_ERR(BT_AICS_ERR_INVALID_COUNTER);
+	}
+
+	switch (cp->cp.opcode) {
+	case BT_AICS_OPCODE_SET_GAIN:
+		BT_DBG("Set gain %d", cp->gain_setting);
+		if (cp->gain_setting < inst->gain_settings.minimum ||
+		    cp->gain_setting > inst->gain_settings.maximum) {
+			return BT_GATT_ERR(BT_AICS_ERR_OUT_OF_RANGE);
+		}
+		if (BT_AICS_INPUT_MODE_SETTABLE(inst->state.gain_mode) &&
+		    inst->state.gain != cp->gain_setting) {
+			inst->state.gain = cp->gain_setting;
+			notify = true;
+		}
+		break;
+	case BT_AICS_OPCODE_UNMUTE:
+		BT_DBG("Unmute");
+		if (inst->state.mute == BT_AICS_STATE_MUTE_DISABLED) {
+			return BT_GATT_ERR(BT_AICS_ERR_MUTE_DISABLED);
+		}
+		if (inst->state.mute != BT_AICS_STATE_UNMUTED) {
+			inst->state.mute = BT_AICS_STATE_UNMUTED;
+			notify = true;
+		}
+		break;
+	case BT_AICS_OPCODE_MUTE:
+		BT_DBG("Mute");
+		if (inst->state.mute == BT_AICS_STATE_MUTE_DISABLED) {
+			return BT_GATT_ERR(BT_AICS_ERR_MUTE_DISABLED);
+		}
+		if (inst->state.mute != BT_AICS_STATE_MUTED) {
+			inst->state.mute = BT_AICS_STATE_MUTED;
+			notify = true;
+		}
+		break;
+	case BT_AICS_OPCODE_SET_MANUAL:
+		BT_DBG("Set manual mode");
+		if (BT_AICS_INPUT_MODE_IMMUTABLE(inst->state.gain_mode)) {
+			return BT_GATT_ERR(BT_AICS_ERR_GAIN_MODE_NOT_ALLOWED);
+		}
+		if (inst->state.gain_mode != BT_AICS_MODE_MANUAL) {
+			inst->state.gain_mode = BT_AICS_MODE_MANUAL;
+			notify = true;
+		}
+		break;
+	case BT_AICS_OPCODE_SET_AUTO:
+		BT_DBG("Set automatic mode");
+		if (BT_AICS_INPUT_MODE_IMMUTABLE(inst->state.gain_mode)) {
+			return BT_GATT_ERR(BT_AICS_ERR_GAIN_MODE_NOT_ALLOWED);
+		}
+		if (inst->state.gain_mode != BT_AICS_MODE_AUTO) {
+			inst->state.gain_mode = BT_AICS_MODE_AUTO;
+			notify = true;
+		}
+		break;
+	default:
+		return BT_GATT_ERR(BT_AICS_ERR_OP_NOT_SUPPORTED);
+	}
+
+	if (notify) {
+		inst->state.change_counter++;
+
+		BT_DBG("New state: gain %d, mute %u, gain_mode %u, counter %u",
+		       inst->state.gain, inst->state.mute,
+		       inst->state.gain_mode, inst->state.change_counter);
+
+		bt_gatt_notify_uuid(NULL, BT_UUID_AICS_STATE,
+				    inst->service_p->attrs, &inst->state,
+				    sizeof(inst->state));
+
+		if (inst->cb && inst->cb->state) {
+			inst->cb->state(NULL, (struct bt_aics *)inst, 0,
+					inst->state.gain, inst->state.mute,
+					inst->state.gain_mode);
```

Comment:
If I get how this is working the new states are notified after the operation which can never fail, while this might work for testing Id expect the entity handling these settings would have to be requested to accept the operation only then we can confirm the operation has been performed and if there is an error report back.

Reply:
Hmm, if I understand you correctly, you'd expect the AICS to request the change towards the application, rather than telling the application that the change was made? 

That's an interesting alternative. I don't really have a strong opinion about this. The request for change is how it's implemented in e.g. OTS (see ots.c:192), but other characteristics such as the device name are silently accepted without the application's approval (see gatt.c:111). 

So basically we have two different ways of handling writes from the client:
1) Ask the application for permission to finalize the write and either accept or reject it
2) Tell the application that a write has happened. 

Currently we seem to employ both in different services. 

I think this is an interesting discussion to have, and I'd like to bring in @asbjornsabo , @jhedberg  and @joerchan as well, although I think this is outside the scope of this particular PR.

Reply:
Set up a discussion for it.
This is a larger question than this PR.

Reply:
Moved to https://github.com/zephyrproject-rtos/zephyr/discussions/33592#discussioncomment-589560